python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -0,0 +1,116 @@
+from typing import TYPE_CHECKING, Dict, List, Union, Iterable
+import struct
+
+from ray.util.annotations import PublicAPI
+from ray.data.block import Block
+from ray.data.datasource.file_based_datasource import FileBasedDatasource
+
+if TYPE_CHECKING:
+    import pyarrow
+    import tensorflow as tf
+
+
+@PublicAPI(stability="alpha")
+class TFRecordDatasource(FileBasedDatasource):
+
+    _FILE_EXTENSION = "tfrecords"
+
+    def _read_file(self, f: "pyarrow.NativeFile", path: str, **reader_args) -> Block:
+        from google.protobuf.message import DecodeError
+        import pandas as pd
+        import tensorflow as tf
+
+        data = []
+        for record in _read_records(f):
+            example = tf.train.Example()
+            try:
+                example.ParseFromString(record)
+            except DecodeError:
+                raise ValueError(
+                    "`TFRecordDatasource` failed to parse `tf.train.Example` "
+                    f"record in '{path}'. This error can occur if your TFRecord "
+                    "file contains a message type other than `tf.train.Example`."
+                )
```

Comment:
Nit: We should include the decoding error at the end of the error message in case it's relevant.
```suggestion
            except DecodeError as e:
                raise ValueError(
                    "`TFRecordDatasource` failed to parse `tf.train.Example` "
                    f"record in '{path}'. This error can occur if your TFRecord "
                    f"file contains a message type other than `tf.train.Example`: {e}"
                )
```